### ISSUE TITLE
Escaping headings

### ISSUE THREAD

**Issue by melisgl** [3b/3bmd#58]
A have a bit of headache with the parse/print consistency of headings.

First, and this may be how markdown works, if there is no newline after the "heading", then it's parsed as `:PLAIN`:

```
CL-USER> (3bmd::parse-doc "x
#y
")
((:PARAGRAPH "x") (:HEADING :LEVEL 1 :CONTENTS ("y")))
NIL
T
CL-USER> (3bmd::parse-doc "x
#y")
((:PARAGRAPH "x") (:PLAIN "#" "y"))
NIL
T
```

When the latter is printed, an extra newline is inserted:

```
CL-USER> (3bmd:print-doc-to-stream (3bmd::parse-doc "x
#y") t :format :markdown)
x

#y
NIL
```

When the heading is escaped, the parse is good, but printing loses the escape:

```
CL-USER> (3bmd::parse-doc "x
\\#y
")
((:PLAIN "x" "
"
  "#" "y"))
NIL
T
CL-USER> (3bmd:print-doc-to-stream (3bmd::parse-doc "x
\\#y
") t)
x
#y
NIL
```

If this output is parsed again, then we get a `:HEADING`. Thus print/parse consistency is lost.

The quick fix would be to escape all # characters in `print-md-escaped`, but that produces unnecessarily cluttered output, which goes against the spirit of markdown. The right solution seems to be to escape only in column 0, but that's not easily and portably available.

**Comment by melisgl** [comment]
Also, is this intended?

```
CL-USER> (3bmd::parse-doc "- xxx

    # yyy
")
((:BULLET-LIST
  (:LIST-ITEM (:PARAGRAPH "xxx") (:HEADING :LEVEL 1 :CONTENTS ("yyy")))))
NIL
T
```

**Comment by 3b** [comment]
newline after heading probably shouldn't be required, should just need to add `EOF` to the rule there.

for the version that didn't parse as a heading, it probably should have included `(:plain ...)` in the `(:paragraph ...)`. In that case it would probably output as `x #y` on one line, which i think would be reasonable.

Escaping nicely is harder, since as you noted in the comment, headings (and all the blocks) can be included in lists and blockquotes, so might start start at arbitrary positions. I think setting a flag in `md-indent` and checking (and then immediately clearing) that in `print-md-escaped` might work?

Probably factor out the set of characters that need escaped for inlines (escape anywhere) and blocks (escape only after `md-indent`) as well, so extensions could add to them (assuming the existing extension hooks don't already give us enough info, not sure from a quick look at the code).

**Comment by 3b** [comment]
Other option for escaping is to store it when parsing, though that would probably complicate the other outputs.

**Comment by 3b** [comment]
fixed along with #60 i think
